Doc to help determine action when CVE is reported

mthalman opened the pull request:
Related to https://github.com/dotnet/dotnet-docker/issues/2787

MichaelSimons reviewed (commented):
I like the direction you are headed @mthalman.  I realize this is still a draft but still felt it was worthwhile to share some thoughts/comments I had.

mthalman commented on documentation/vulnerability-reporting.md:
````diff
@@ -0,0 +1,239 @@
+1. Run the following command:
+
+    ```shell
+    docker pull <insert-my-image-digest>
````
I use that in the other script because that's specific to Linux containers. Your suggestion will fail if attempting to pull an image of a different OS type. So you may not even get to the point of running `docker inspect`. That's why the instructions here make the determination based on the error message or not.

mthalman commented on documentation/vulnerability-reporting.md:
````diff
@@ -0,0 +1,239 @@
+By answering each question, it will eventually direct you to the action to help resolve your vulnerability.
+
+```mermaid
+flowchart
````
I would say no. You need a magnifying glass with horizontal.

mthalman commented on documentation/vulnerability-reporting.md:
````diff
@@ -0,0 +1,239 @@
+#### Linux
+
+```shell
+docker manifest inspect <image-tag> | jq -r .mediaType
````
Experimental mode is [enabled by default](https://docs.docker.com/engine/reference/commandline/cli/#experimental-features) now. In fact, you can't even disable it: https://github.com/docker/cli/commit/977d3ae046ec6c64be8788a8712251ed547a2bdb. For that reason, I don't think it's worth mentioning.

mthalman commented on documentation/vulnerability-reporting.md:
````diff
@@ -0,0 +1,239 @@
+
+```shell
+myImage="<insert-my-image-digest>"
+baseImage="<insert-base-image-tag>"
````
Good suggestion. That was actually a very easy script to write.

MichaelSimons commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,399 @@
+
+#### Linux
+
+Requires [PowerShell](https://learn.microsoft.com/powershell/scripting/install/installing-powershell-on-linux) and [Syft](https://github.com/anchore/syft) to be installed.
```
Given the dependency on an external tool (e.g. syft), consider describing in a bit more detail how this can be verified manually.

mthalman replied:
The whole intent of using Syft was to avoid describing in detail how to get this information. It's specific to each package manager type and distroless is a whole other mess to deal with. If the concern is about requiring the downloading of a 3rd party tool, I could change the script to make use of Syft's Docker image instead. Would that be better?

MichaelSimons replied:
Yes, I think having a docker based solution would address my primary concern.

MichaelSimons commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,399 @@
+
+#### Linux
+
+Requires [PowerShell to be installed](https://learn.microsoft.com/powershell/scripting/install/installing-powershell-on-linux).
```
The Syft dependency should be mentioned here in case the reader skips to this step.

MichaelSimons commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,399 @@
+### Q. Log an issue at dotnet/dotnet-docker
+
+This vulnerability can be directly addressed by the .NET team.
+Log an issue at [dotnet/dotnet-docker](https://github.com/dotnet/dotnet-docker) and provide the vulnerability ID and the .NET image tag you've identified as being affected.
```
Consider linking directly to the create issue page for convienence.

MichaelSimons commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,399 @@
+
+#### Linux
+
+Requires [PowerShell to be installed](https://learn.microsoft.com/powershell/scripting/install/installing-powershell-on-linux).
```
This is a PIA.  ...but I agree with the decision to get this out there to gauge value before investing further.

MichaelSimons reviewed (commented):
Looks great, I can wait to get this merged so we can start directing users to it.

mthalman commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,399 @@
+    PlatformType --> |Linux|LinuxFixAvailability{H. Is fix available<br>for the package<br>from Linux distro?}
+    PlatformType --> |Windows|WindowsVulnSource{O. What is the<br>source of the<br>vulnerability?}
+    LinuxFixAvailability --> |No|WaitForFix[I. Wait for fix to be available<br>from Linux distro]
+    LinuxFixAvailability --> |Yes|InLinuxBaseImage{J. Is package in<br>Linux distro<br>base image?}
```
I originally did that. But it led to a messy graph with lines crossing. I intentionally switched the yes/no sides for certain cases to optimize the placements of the boxes to produce a clean look.

mthalman commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,399 @@
+
+Most image scanning software will output the [image digest](#what-is-an-image-digest) of the image that is being scanned in its report or logs.
+Be sure that that digest represents the latest version of your image.
+To get the digest of your image, be sure to target a tag that is not a multi-arch/manifest list tag.
```
I will update this to be more informative on how to resolve the multi-arch tag. For your comment about pulling, that's specified in the next section: "Scan your latest image". This section isn't about what action to take. It's about determining the current state.

mthalman commented on documentation/vulnerability-workflow.md:
````diff
@@ -0,0 +1,399 @@
+
+```shell
+docker pull <image-name>
+docker run --rm --entrypoint /bin/sh <image-name> -c "apt update > /dev/null && apt list --upgradable"
````
True but vulnerability scanners aren't currently going to find any vulnerabilities in chiseled images anyway right now. We'll have to come back to this once there is scanner support and that should help us determine a way forward on this.

MichaelSimons replied:
Agreed, sounds reasonable.

blowdart commented on documentation/vulnerability-workflow.md:
````diff
@@ -0,0 +1,429 @@
+$myImage="<insert-my-image-tag>"
+[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
+&([scriptblock]::Create((Invoke-WebRequest -UseBasicParsing 'https://raw.githubusercontent.com/dotnet/dotnet-docker/main/documentation/scripts/resolve-image-digest.ps1'))) $myImage
+```
````
It's always fun to see security documentation recommending that you download and execute random files from github :)

Perhaps provide instructions for saving the ps1 files from the repo to the local drive too?

mthalman replied:
> Perhaps provide instructions for saving the ps1 files from the repo to the local drive too?

Is that really necessary? If someone doesn't want to execute the script as is, surely they can be relied upon to know how to get the contents of the clearly visible URL onto their local machine.

blowdart commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,429 @@
+
+Switching to a supported base image tag requires knowing the specific tag usage:
+
+* If you're using an out-of-support .NET version, such as .NET 5, you'll need to upgrade your project and Dockerfiles to target a supported version.
```
Nit: need to => must. Stronger language is preferred.

Same for the points.

blowdart commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,429 @@
+* If you're using an out-of-support .NET version, such as .NET 5, you'll need to upgrade your project and Dockerfiles to target a supported version.
+* If you're using a patch version tag which which has been replaced by a newer patch version, you'll need to switch to the latest patch version. For example, `6.0.8` => `6.0.9`
+* If you're using a version-specific Alpine Linux tag for a version of Alpine that is no longer being maintained via .NET container images, you'll need to move to the latest version. For example, `6.0-alpine3.15` => `6.0-alpine3.16`
+* If you're using a tag for a specific Windows version that is no longer supported, you'll need to move to a supported version of Windows. For example, `6.0-nanoserver-20H2` => `6.0-nanoserver-ltsc2022`
```
By adding specific version numbers here you're committing to updating this every month as 6.0 gets a new patch release, otherwise someone will point to this and say "This example says 6.0.9 is good". I realise this may seem unreasonable, but I've seen it happen.

blowdart commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,429 @@
+Run one of the scripts below to determine whether your image is built on the latest image for that tag. Provide the [image digest](#what-is-an-image-digest) of your image and tag of the .NET base image (e.g. `mcr.microsoft.com/dotnet/aspnet:6.0`) as input.
+
+A result of `True` means that your image is built from the latest .NET base image.
+A result of `False` means it is not built on the latest.
```
not built on the latest _and must be updated_

Yes, it's repetitive, but that's on purpose. People will skin to what to I run and what do the results mean.

blowdart commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,429 @@
+### F. Rebuild your image to get latest base
+
+Since your image is out of date with the latest version of the .NET base image, rebuilding it will ensure it's up-to-date.
+Be sure that your workflow explicitly pulls the .NET base image tag rather than using a potentially cached version of it.
```
How? Cover azdo and actions perhaps?

blowdart commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,429 @@
+The severity of a vulnerability is indicated by its Common Vulnerability Scoring System (CVSS) base score.
+To find this score, search for the CVE at the [CVE site](https://www.cve.org/SiteSearch).
+The score will consist of both a number and a label.
+Note whether the label indicates "High" or "Critical".
```
Not everyone uses CVSS unfortunately

mthalman replied:
In that case then I doubt we'd get anyone coming to this repo about it then. All of the requests that I've ever received about vulnerabilities have been classified as high or critical.

blowdart commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,429 @@
+
+### O. What is the source of the vulnerability?
+
+The title or description of the vulnerability should indicate whether it's for .NET or not.
```
Bet :) It may be useful to point out CPEs, bugs in Microsoft products will always begin with

cpe:2.3:a:microsoft or, using the older CPE format  cpe:/a:microsoft

mthalman replied:
Given that this question is for the Windows path, all vulnerabilities would be Microsoft-related. This is specifically trying to narrow that down to with it is .NET-related or not.

blowdart commented on documentation/vulnerability-workflow.md:
```diff
@@ -0,0 +1,429 @@
+### P. Wait for the next .NET servicing release
+
+Given that the vulnerability does not have a high severity, it will be addressed in the next .NET servicing release.
+These occur on Patch Tuesday (the second Tuesday of each month).
```
Officially it's Update Tuesday :)

blowdart reviewed (changes requested):
You should also talk to Rich, given https://devblogs.microsoft.com/dotnet/investigating-a-linux-cve-with-net-images/

MichaelSimons commented on documentation/scripts/check-latest-base.ps1:
```diff
@@ -0,0 +1,86 @@
+    # The image to be checked on whether it is based on the latest version of a .NET base image tag.
+    [Parameter(Mandatory = $true)]
+    [string]
+    $ImageTag,
```
I see a number of unresolved comments from my last review.  Not sure if they were overlooked or something you had a different view on.  I'm going to hold off on reviewing until I see a response.

mthalman commented:
I believe I've addressed all remaining comments.

In addition, I made an update to the flowchart to handle the scenario of a scanning tool reporting on a vulnerability for Linux. This was accounted for in the Windows path of the flowchart but not the Linux path. I've updated the chart to handle this. Linux is notable in this case because our images don't have Linux packages for .NET installed, we just have the extracted tarball which is what the scanning tools are looking for. So we need to handle that in the workflow before getting into the questioning about Linux packages.

MichaelSimons commented on documentation/vulnerability-reporting.md:
````diff
@@ -0,0 +1,447 @@
+```shell
+myImage="<insert-my-image-digest>" # format: <registry>/<repo>@sha256:<digest>
+dotnetBaseImage="<insert-dotnet-image-tag>" # example: mcr.microsoft.com/dotnet/aspnet:7.0
+curl -sSL https://raw.githubusercontent.com/dotnet/dotnet-docker/main/documentation/scripts/check-latest-bash.ps1 | pwsh /dev/stdin $myImage $dotnetBaseImage
````
```suggestion
curl -sSL https://raw.githubusercontent.com/dotnet/dotnet-docker/main/documentation/scripts/check-latest-base.ps1 | pwsh /dev/stdin $myImage $dotnetBaseImage
```

MichaelSimons commented on documentation/vulnerability-reporting.md:
```diff
@@ -0,0 +1,447 @@
+$myImage="<insert-my-image-digest>" # format: <registry>/<repo>@sha256:<digest>
+$dotnetBaseImage="<insert-dotnet-image-tag>" # example: mcr.microsoft.com/dotnet/aspnet:7.0
+[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
+&([scriptblock]::Create((Invoke-WebRequest -UseBasicParsing 'https://raw.githubusercontent.com/dotnet/dotnet-docker/main/documentation/scripts/check-latest-bash.ps1'))) $myImage $dotnetBaseImage
```
```suggestion
&([scriptblock]::Create((Invoke-WebRequest -UseBasicParsing 'https://raw.githubusercontent.com/dotnet/dotnet-docker/main/documentation/scripts/check-latest-base.ps1'))) $myImage $dotnetBaseImage
```

MichaelSimons commented on documentation/vulnerability-reporting.md:
````diff
@@ -0,0 +1,447 @@
+
+#### Windows
+
+```powershell
````
I'm getting an error when executing this:

```
Exception calling "Create" with "1" argument(s): "At line:58 char:9
+         | Where-Object { $_.platform.Architecture -eq $imageConfig.Ar ...
+         ~
An empty pipe element is not allowed.
At line:59 char:9
+         | Select-Object -ExpandProperty digest
+         ~
An empty pipe element is not allowed."
At line:1 char:1
+ &([scriptblock]::Create((Invoke-WebRequest -UseBasicParsing 'https:// ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : ParseException
```

Here is the input I ran:

```
$myImage="mcr.microsoft.com/dotnet/samples@sha256:6274b2bdaba286678f4873a56194ea310f404ec9ae763e27a3d1a65f12ec5d26" # format: <registry>/<repo>@sha256:<digest>
$dotnetBaseImage="mcr.microsoft.com/dotnet/runtime-deps:7.0-alpine-amd64" # example: mcr.microsoft.com/dotnet/aspnet:7.0
[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
&([scriptblock]::Create((Invoke-WebRequest -UseBasicParsing 'https://raw.githubusercontent.com/dotnet/dotnet-docker/d18e5659ec1f95735a1706c76b69cfd71c4ffc86/documentation/scripts/check-latest-base.ps1'))) $myImage $dotnetBaseImage
```

mthalman replied:
The digest you specified doesn't exist:

```
Pulling mcr.microsoft.com/dotnet/samples@sha256:6274b2bdaba286678f4873a56194ea310f404ec9ae763e27a3d1a65f12ec5d26...
Error response from daemon: manifest for mcr.microsoft.com/dotnet/samples@sha256:6274b2bdaba286678f4873a56194ea310f404ec9ae763e27a3d1a65f12ec5d26 not found: manifest unknown: manifest sha256:6274b2bdaba286678f4873a56194ea310f404ec9ae763e27a3d1a65f12ec5d26 is not found
```

Can you provide a valid one? I'm unable to reproduce what you're seeing with the ones I've tried.

MichaelSimons replied:
I am still getting the same error.  This time I validated the sha was valid.

``` bash
> docker pull mcr.microsoft.com/dotnet/samples@sha256:e59ce8b5c10e0676f60954fab0494fcc27d845bd15bf08ccde996305f44fadb7
mcr.microsoft.com/dotnet/samples@sha256:e59ce8b5c10e0676f60954fab0494fcc27d845bd15bf08ccde996305f44fadb7: Pulling from dotnet/samples
Digest: sha256:e59ce8b5c10e0676f60954fab0494fcc27d845bd15bf08ccde996305f44fadb7
Status: Downloaded newer image for mcr.microsoft.com/dotnet/samples@sha256:e59ce8b5c10e0676f60954fab0494fcc27d845bd15bf08ccde996305f44fadb7
mcr.microsoft.com/dotnet/samples@sha256:e59ce8b5c10e0676f60954fab0494fcc27d845bd15bf08ccde996305f44fadb7
```

``` bash
PS C:\Users\msimons\Downloads> $myImage="mcr.microsoft.com/dotnet/samples@sha256:e59ce8b5c10e0676f60954fab0494fcc27d845bd15bf08ccde996305f44fadb7" # format: <registry>/<repo>@sha256:<digest>
PS C:\Users\msimons\Downloads> $dotnetBaseImage="mcr.microsoft.com/dotnet/runtime-deps:7.0-alpine-amd64" # example: mcr.microsoft.com/dotnet/aspnet:7.0
PS C:\Users\msimons\Downloads> [Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
PS C:\Users\msimons\Downloads> &([scriptblock]::Create((Invoke-WebRequest -UseBasicParsing 'https://raw.githubusercontent.com/dotnet/dotnet-docker/d18e5659ec1f95735a1706c76b69cfd71c4ffc86/documentation/scripts/check-latest-base.ps1'))) $myImage $dotnetBaseImage
Exception calling "Create" with "1" argument(s): "At line:58 char:9
+         | Where-Object { $_.platform.Architecture -eq $imageConfig.Ar ...
+         ~
An empty pipe element is not allowed.
At line:59 char:9
+         | Select-Object -ExpandProperty digest
+         ~
An empty pipe element is not allowed."
At line:1 char:1
+ &([scriptblock]::Create((Invoke-WebRequest -UseBasicParsing 'https:// ...
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [], MethodInvocationException
    + FullyQualifiedErrorId : ParseException
```

MichaelSimons commented on documentation/vulnerability-reporting.md:
````diff
@@ -0,0 +1,447 @@
+Requires [PowerShell](https://learn.microsoft.com/powershell/scripting/install/installing-powershell-on-linux) to be installed.
+
+```shell
+imageName="<insert-linux-base-image-name>"
````
Consider adding an example here e.g. # example: alpine:3.17

MichaelSimons commented on documentation/vulnerability-reporting.md:
````diff
@@ -0,0 +1,447 @@
+
+```shell
+imageName="<insert-linux-base-image-name>"
+curl -sSL https://raw.githubusercontent.com/dotnet/dotnet-docker/main/documentation/scripts/check-package-install.ps1 | pwsh /dev/stdin $imageName
````
Isn't this missing the package name?  Same applies to the windows snippet.

MichaelSimons replied:
It would also be helpful to clarify the format of the package name e.g. the version numbers can be confusing to people who are not familiar with packages.

mthalman replied:
This is added in section J.

MichaelSimons commented on documentation/vulnerability-reporting.md:
````diff
@@ -0,0 +1,447 @@
+
+```shell
+imageName="<insert-dotnet-image-name>" # example: mcr.microsoft.com/dotnet/aspnet:7.0
+curl -sSL https://raw.githubusercontent.com/dotnet/dotnet-docker/main/documentation/scripts/check-package-install.ps1 | pwsh /dev/stdin $imageName
````
Same issue here as earlier, the package name option is missing

MichaelSimons commented on documentation/vulnerability-reporting.md:
```diff
@@ -0,0 +1,447 @@
+### O. Rebuild your image to get the latest package
+
+Because the package is not installed by the Linux distro base image or .NET, it must be installed by your Dockerfile.
+Check your Dockerfile to check how the package is being installed.
```
Avoiding using `Check` multiple times in this sentence.

MichaelSimons commented on documentation/vulnerability-reporting.md:
```diff
@@ -0,0 +1,447 @@
+
+### Q. Wait for the next .NET servicing release
+
+Given that the vulnerability does not have a high severity, it will be addressed in the next .NET servicing release.
```
```suggestion
Given that the vulnerability does not have a critical or high severity, it will be addressed in the next .NET servicing release.
```

MichaelSimons commented on documentation/vulnerability-reporting.md:
```diff
@@ -0,0 +1,447 @@
+
+An image digest is a unique identifier for the image.
+It allows you to retrieve that specific image and know that you're getting the same version as someone else that also used that same image digest.
+Contrast image digests with image tags which are mutable and cannot be relied upon for ensuring you're getting an exact version of the image every time.
```
This reads a little strange to me.

```suggestion
Contrasting image tags with image digests, image tags are mutable and cannot be relied upon for ensuring you're getting an exact version of the image every time.
```

MichaelSimons commented on documentation/vulnerability-reporting.md:
```diff
@@ -0,0 +1,447 @@
+
+### How can I determine whether my image is based on Linux or Windows?
+
+1. If you're using Docker Desktop for Windows, [change your settings](https://docs.docker.com/desktop/settings/windows/) to target Linux containers.
```
It will be interesting to see if we get feedback here from MacOS users.

mthalman replied:
It shouldn't matter. The only cases you need to actually run containers locally is when you're investigating a vulnerability in a Linux container. So the instructions are valid for all OS types.

mthalman commented on documentation/vulnerability-reporting.md:
````diff
@@ -0,0 +1,447 @@
+```powershell
+dotnetImage="<insert-dotnet-image-tag>" # example: mcr.microsoft.com/dotnet/aspnet:7.0
+[Net.ServicePointManager]::SecurityProtocol = [Net.SecurityProtocolType]::Tls12
+&([scriptblock]::Create((Invoke-WebRequest -UseBasicParsing 'https://raw.githubusercontent.com/dotnet/dotnet-docker/main/documentation/scripts/check-tag-support.ps1'))) $dotnetImage
````
This was because I added strict mode after testing it. Doh. Fixed.

MichaelSimons commented on documentation/vulnerability-reporting.md:
Consider referencing the distro's package manager here.

```suggestion
But this indicates the name of the source package, not the name of the binary package that would be installed from the distro's package manager.
```
